1. Basics/task_4/hw4.php

```diff
@@ -0,0 +1,315 @@
+ * 
+ * @param array $arr
+ * 
+ * @return array
+ */
+function cleanArray($arr) : array {
+    $resultArr = [];
+    
+    foreach($arr as $item) {
+        if(strlen($item) > 0){
+            $resultArr[]=$item;
+        }
+    }
+
+    return $resultArr;
+}
+
+/**
+ * @param string $string
+ * @param string $arrLastItem
+ * 
+ * @return bool
+ */
+function hasBody($string, $arrLastItem) : bool {
+    return preg_match('/Content-Length:/', $string) || !preg_match('/:/', $arrLastItem);
+}
+
+/**
+ * Analyzes request anr return status code for response
+ * 
+ * @param string    $method
+ * @param string    $uri
+ * @param string    $contentType
+ * @param array     $body
+ * 
+ * @return int
+ */
+function getResponseStatusCode($method, $uri, $contentType, $body) : int {
+    
+    // 400 : Bad Request (invalid method)
+    if($method != "POST") return 400;
+
+    // 400 : Bad Request (invalid "Content-type" header)
+    if($contentType != "application/x-www-form-urlencoded" ) return 400;
+
+    // 400 : Bad Request (invalid body)
+    if(!isRequestBodyValid($body)) return 400;
+
+    // 404 : Not Found (invalid uri)
+    if($uri != "/api/checkLoginAndPassword") return 404;
+    
+    // options: 200, 500, 404
+    return getAccessStatus($body);
+}
+
+/**
+ * Checks if array $body contains "login" and "password" as its keys.
+ * 
+ * @param array $body
+ * 
+ * @return bool 
+ */
+function isRequestBodyValid($body) {
+    return (array_key_exists("login", $body) && array_key_exists("password", $body));
+}
+
+/**
+ * Returns one of 200, 404, 500
+ * 
+ * @param array $body
+ * 
+ * @return int
+ */
+function getAccessStatus($body) : int {
+
+    // 500 : Internal Server Error (file not exist)
+    if(!file_exists(FILE_NAME)) return 500;
+
+    $logAndPassPairsAsArr = getFileContentAsPairsArray(FILE_NAME, PHP_EOL, ":");
+
+    foreach($logAndPassPairsAsArr as $login => $password) {
+        
+        if($body['login'] == $login) {
+            if($body['password'] == $password) {
+                
+                // login found, password matched
+                return 200;
+            }
+
+            // login found, but password DIDN'T match
+            else        
+                break;
+        }
+    }
+
+    // 404 : Not Found (login or/and password not found)
+    return 404;
+}
+
+/**
+ * @param string $filename
+ * @param string $delim1
+ * @param string $delim2
+ * 
+ * @return array
+ */
+function getFileContentAsPairsArray($filename, $delim1, $delim2) : array {
+    return getStringContentAsPairsArray(file_get_contents($filename), $delim1, $delim2);
+}
+
+/**
+ * @param string $str
+ * @param string $delim1
+ * @param string $delim2
+ * 
+ * @return array
+ */
+function getStringContentAsPairsArray($str, $delim1, $delim2) : array {
+    $arr = cleanArray(explode($delim1, $str));
+    $pairsAsArr = [];
+    
+    foreach($arr as $pair) {
+        $p = explode($delim2, $pair);subject: 
```

Comment:
what is `subject:`?

Reply:
![image](https://github.com/user-attachments/assets/5a7410c0-cfcc-497c-94ef-0c2d1f0c5dd0)
Its color tells that is the part on my new debugger using result. Now I have no ideas how it became the part of my code.
I'll remove it.